File: binding/src/libqjsonarray.cpp
#include <QDebug>
#include <QJsonArray>
#define WORKAROUND_INNER_CLASS_DEFINITION_QJsonArray__const_iterator
#define WORKAROUND_INNER_CLASS_DEFINITION_QJsonArray__iterator
#include <QJsonValue>
#include <QRect>
#include <qjsonarray.h>
#include "libqjsonarray.h"
#include "libqjsonarray.hxx"

QJsonArray* QJsonArray_new() {
	 return new QJsonArray();
}

QJsonArray* QJsonArray_new2(const QJsonArray* other) {
	 return new QJsonArray(*other);
}

void QJsonArray_OperatorAssign(QJsonArray* self, const QJsonArray* other) {
	self->operator=(*other);
}

QJsonArray* QJsonArray_FromStringList(const libqt_list list) {
	return new QJsonArray(QJsonArray::fromStringList(QList<QString>()));
}

QJsonArray* QJsonArray_FromVariantList(const libqt_list list) {
	return new QJsonArray(QJsonArray::fromVariantList(QList<QVariant>()));
}

libqt_list QJsonArray_ToVariantList(const QJsonArray* self) {
	auto _ret = self->toVariantList();
	libqt_list _arr;
	_arr.len = _ret.length();
	_arr.data = malloc(_arr.len * sizeof(void*));
	void** _data = static_cast<void**>(_arr.data);
	for (int _i = 0; _i < _arr.len; ++_i) {
		auto& _elem = _ret[_i];
		_data[_i] = new std::remove_reference_t<decltype(_elem)>(_elem);
	}
	return _arr;
}

ptrdiff_t QJsonArray_Size(const QJsonArray* self) {
	return self->size();
}

ptrdiff_t QJsonArray_Count(const QJsonArray* self) {
	return self->count();
}

bool QJsonArray_IsEmpty(const QJsonArray* self) {
	return self->isEmpty();
}

QJsonValue* QJsonArray_At(const QJsonArray* self, ptrdiff_t i) {
	return new QJsonValue(self->at(i));
}

QJsonValue* QJsonArray_First(const QJsonArray* self) {
	return new QJsonValue(self->first());
}

QJsonValue* QJsonArray_Last(const QJsonArray* self) {
	return new QJsonValue(self->last());
}

void QJsonArray_Prepend(QJsonArray* self, const QJsonValue* value) {
	self->prepend(*value);
}

void QJsonArray_Append(QJsonArray* self, const QJsonValue* value) {
	self->append(*value);
}

void QJsonArray_RemoveAt(QJsonArray* self, ptrdiff_t i) {
	self->removeAt(i);
}

QJsonValue* QJsonArray_TakeAt(QJsonArray* self, ptrdiff_t i) {
	return new QJsonValue(self->takeAt(i));
}

void QJsonArray_RemoveFirst(QJsonArray* self) {
	self->removeFirst();
}

void QJsonArray_RemoveLast(QJsonArray* self) {
	self->removeLast();
}

void QJsonArray_Insert(QJsonArray* self, ptrdiff_t i, const QJsonValue* value) {
	self->insert(i, *value);
}

void QJsonArray_Replace(QJsonArray* self, ptrdiff_t i, const QJsonValue* value) {
	self->replace(i, *value);
}

bool QJsonArray_Contains(const QJsonArray* self, const QJsonValue* element) {
	return self->contains(*element);
}

QJsonValue* QJsonArray_OperatorSubscript2(const QJsonArray* self, ptrdiff_t i) {
	return new QJsonValue(self->operator[](i));
}

void QJsonArray_Swap(QJsonArray* self, QJsonArray* other) {
	self->swap(*other);
}

It::value_type* QJsonArray_Begin(QJsonArray* self) {
	return self->begin();
}

const QRect* QJsonArray_Begin2(const QJsonArray* self) {
	return self->begin();
}

const QRect* QJsonArray_ConstBegin(const QJsonArray* self) {
	return self->constBegin();
}

const QRect* QJsonArray_Cbegin(const QJsonArray* self) {
	return self->cbegin();
}

It::value_type* QJsonArray_End(QJsonArray* self) {
	return self->end();
}

const QRect* QJsonArray_End2(const QJsonArray* self) {
	return self->end();
}

const QRect* QJsonArray_ConstEnd(const QJsonArray* self) {
	return self->constEnd();
}

const QRect* QJsonArray_Cend(const QJsonArray* self) {
	return self->cend();
}

It::value_type* QJsonArray_Insert2(QJsonArray* self, It::value_type* before, const QJsonValue* value) {
	return self->insert(before, *value);
}

It::value_type* QJsonArray_Erase(QJsonArray* self, It::value_type* it) {
	return self->erase(it);
}

QJsonArray* QJsonArray_OperatorPlus(const QJsonArray* self, const QJsonValue* v) {
	return new QJsonArray(self->operator+(*v));
}

QJsonArray* QJsonArray_OperatorPlusAssign(QJsonArray* self, const QJsonValue* v) {
	return new QJsonArray(self->operator+=(*v));
}

QJsonArray* QJsonArray_OperatorShiftLeft(QJsonArray* self, const QJsonValue* v) {
	return new QJsonArray(self->operator<<(*v));
}

void QJsonArray_PushBack(QJsonArray* self, const QJsonValue* t) {
	self->push_back(*t);
}

void QJsonArray_PushFront(QJsonArray* self, const QJsonValue* t) {
	self->push_front(*t);
}

void QJsonArray_PopFront(QJsonArray* self) {
	self->pop_front();
}

void QJsonArray_PopBack(QJsonArray* self) {
	self->pop_back();
}

bool QJsonArray_Empty(const QJsonArray* self) {
	return self->empty();
}

void QJsonArray_Delete(QJsonArray* self) {
    delete self;
}

QJsonArray__iterator* QJsonArray__iterator_new() {
	 return new QJsonArray::iterator();
}

QJsonArray__iterator* QJsonArray__iterator_new2(QJsonArray* array, ptrdiff_t index) {
	 return new QJsonArray::iterator(array, index);
}

It::value_type* QJsonArray__iterator_OperatorPlusPlus(QJsonArray__iterator* self) {
	return self->operator++();
}

It::value_type* QJsonArray__iterator_OperatorPlusPlus2(QJsonArray__iterator* self, int param1) {
	return self->operator++(param1);
}

It::value_type* QJsonArray__iterator_OperatorMinusMinus(QJsonArray__iterator* self) {
	return self->operator--();
}

It::value_type* QJsonArray__iterator_OperatorMinusMinus2(QJsonArray__iterator* self, int param1) {
	return self->operator--(param1);
}

It::value_type* QJsonArray__iterator_OperatorPlusAssign(QJsonArray__iterator* self, ptrdiff_t j) {
	return self->operator+=(j);
}

It::value_type* QJsonArray__iterator_OperatorMinusAssign(QJsonArray__iterator* self, ptrdiff_t j) {
	return self->operator-=(j);
}

It::value_type* QJsonArray__iterator_OperatorPlus(const QJsonArray__iterator* self, ptrdiff_t j) {
	return self->operator+(j);
}

It::value_type* QJsonArray__iterator_OperatorMinus(const QJsonArray__iterator* self, ptrdiff_t j) {
	return self->operator-(j);
}

void QJsonArray__iterator_Delete(QJsonArray__iterator* self) {
    delete self;
}

QJsonArray__const_iterator* QJsonArray__const_iterator_new() {
	 return new QJsonArray::const_iterator();
}

QJsonArray__const_iterator* QJsonArray__const_iterator_new2(const QJsonArray* array, ptrdiff_t index) {
	 return new QJsonArray::const_iterator(array, index);
}

QJsonArray__const_iterator* QJsonArray__const_iterator_new3(const It::value_type* o) {
	 return new QJsonArray::const_iterator(o);
}

const QRect* QJsonArray__const_iterator_OperatorPlusPlus(QJsonArray__const_iterator* self) {
	const_iterator& _ret = self->operator++();
	// Cast returned reference into pointer
	return const_cast<QRect*>(&_ret);
}

const QRect* QJsonArray__const_iterator_OperatorPlusPlus2(QJsonArray__const_iterator* self, int param1) {
	return self->operator++(param1);
}

const QRect* QJsonArray__const_iterator_OperatorMinusMinus(QJsonArray__const_iterator* self) {
	const_iterator& _ret = self->operator--();
	// Cast returned reference into pointer
	return const_cast<QRect*>(&_ret);
}

const QRect* QJsonArray__const_iterator_OperatorMinusMinus2(QJsonArray__const_iterator* self, int param1) {
	return self->operator--(param1);
}

const QRect* QJsonArray__const_iterator_OperatorPlusAssign(QJsonArray__const_iterator* self, ptrdiff_t j) {
	const_iterator& _ret = self->operator+=(j);
	// Cast returned reference into pointer
	return const_cast<QRect*>(&_ret);
}

const QRect* QJsonArray__const_iterator_OperatorMinusAssign(QJsonArray__const_iterator* self, ptrdiff_t j) {
	const_iterator& _ret = self->operator-=(j);
	// Cast returned reference into pointer
	return const_cast<QRect*>(&_ret);
}

const QRect* QJsonArray__const_iterator_OperatorPlus(const QJsonArray__const_iterator* self, ptrdiff_t j) {
	return self->operator+(j);
}

const QRect* QJsonArray__const_iterator_OperatorMinus(const QJsonArray__const_iterator* self, ptrdiff_t j) {
	return self->operator-(j);
}

void QJsonArray__const_iterator_Delete(QJsonArray__const_iterator* self) {
    delete self;
}

File: binding/include/libqjsonarray.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qjsonarray.html
pub const qjsonarray = struct {

    /// New constructs a new QJsonArray object.
    pub fn New() QtC.QJsonArray {
        return qtc.QJsonArray_new();
    }


    /// New2 constructs a new QJsonArray object.
    pub fn New2(other: ?*anyopaque) QtC.QJsonArray {
        return qtc.QJsonArray_new2(@ptrCast(other));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QJsonArray_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn FromStringList(list: []const u8) QtC.QJsonArray {
        return qtc.QJsonArray_FromStringList(@ptrCast(list));
    }

    pub fn FromVariantList(list: []const u8) QtC.QJsonArray {
        return qtc.QJsonArray_FromVariantList(@ptrCast(list));
    }

    pub fn ToVariantList(self: ?*anyopaque, ) []const u8 {
        return qtc.QJsonArray_ToVariantList(@ptrCast(self));
    }

    pub fn Size(self: ?*anyopaque, ) i64 {
        return qtc.QJsonArray_Size(@ptrCast(self));
    }

    pub fn Count(self: ?*anyopaque, ) i64 {
        return qtc.QJsonArray_Count(@ptrCast(self));
    }

    pub fn IsEmpty(self: ?*anyopaque, ) bool {
        return qtc.QJsonArray_IsEmpty(@ptrCast(self));
    }

    pub fn At(self: ?*anyopaque, i: i64) QtC.QJsonValue {
        return qtc.QJsonArray_At(@ptrCast(self), i);
    }

    pub fn First(self: ?*anyopaque, ) QtC.QJsonValue {
        return qtc.QJsonArray_First(@ptrCast(self));
    }

    pub fn Last(self: ?*anyopaque, ) QtC.QJsonValue {
        return qtc.QJsonArray_Last(@ptrCast(self));
    }

    pub fn Prepend(self: ?*anyopaque, value: ?*anyopaque) void {
        qtc.QJsonArray_Prepend(@ptrCast(self), @ptrCast(value));
    }

    pub fn Append(self: ?*anyopaque, value: ?*anyopaque) void {
        qtc.QJsonArray_Append(@ptrCast(self), @ptrCast(value));
    }

    pub fn RemoveAt(self: ?*anyopaque, i: i64) void {
        qtc.QJsonArray_RemoveAt(@ptrCast(self), i);
    }

    pub fn TakeAt(self: ?*anyopaque, i: i64) QtC.QJsonValue {
        return qtc.QJsonArray_TakeAt(@ptrCast(self), i);
    }

    pub fn RemoveFirst(self: ?*anyopaque, ) void {
        qtc.QJsonArray_RemoveFirst(@ptrCast(self));
    }

    pub fn RemoveLast(self: ?*anyopaque, ) void {
        qtc.QJsonArray_RemoveLast(@ptrCast(self));
    }

    pub fn Insert(self: ?*anyopaque, i: i64, value: ?*anyopaque) void {
        qtc.QJsonArray_Insert(@ptrCast(self), i, @ptrCast(value));
    }

    pub fn Replace(self: ?*anyopaque, i: i64, value: ?*anyopaque) void {
        qtc.QJsonArray_Replace(@ptrCast(self), i, @ptrCast(value));
    }

    pub fn Contains(self: ?*anyopaque, element: ?*anyopaque) bool {
        return qtc.QJsonArray_Contains(@ptrCast(self), @ptrCast(element));
    }

    pub fn OperatorSubscript2(self: ?*anyopaque, i: i64) QtC.QJsonValue {
        return qtc.QJsonArray_OperatorSubscript2(@ptrCast(self), i);
    }

    pub fn Swap(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QJsonArray_Swap(@ptrCast(self), @ptrCast(other));
    }

    pub fn Begin(self: ?*anyopaque, ) It::value_type {
        return @ptrCast(qtc.QJsonArray_Begin(@ptrCast(self)));
    }

    pub fn Begin2(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonArray_Begin2(@ptrCast(self));
    }

    pub fn ConstBegin(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonArray_ConstBegin(@ptrCast(self));
    }

    pub fn Cbegin(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonArray_Cbegin(@ptrCast(self));
    }

    pub fn End(self: ?*anyopaque, ) It::value_type {
        return @ptrCast(qtc.QJsonArray_End(@ptrCast(self)));
    }

    pub fn End2(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonArray_End2(@ptrCast(self));
    }

    pub fn ConstEnd(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonArray_ConstEnd(@ptrCast(self));
    }

    pub fn Cend(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonArray_Cend(@ptrCast(self));
    }

    pub fn Insert2(self: ?*anyopaque, before: *It::value_type, value: ?*anyopaque) It::value_type {
        return @ptrCast(qtc.QJsonArray_Insert2(@ptrCast(self), @ptrCast(before), @ptrCast(value)));
    }

    pub fn Erase(self: ?*anyopaque, it: *It::value_type) It::value_type {
        return @ptrCast(qtc.QJsonArray_Erase(@ptrCast(self), @ptrCast(it)));
    }

    pub fn OperatorPlus(self: ?*anyopaque, v: ?*anyopaque) QtC.QJsonArray {
        return qtc.QJsonArray_OperatorPlus(@ptrCast(self), @ptrCast(v));
    }

    pub fn OperatorPlusAssign(self: ?*anyopaque, v: ?*anyopaque) QtC.QJsonArray {
        return qtc.QJsonArray_OperatorPlusAssign(@ptrCast(self), @ptrCast(v));
    }

    pub fn OperatorShiftLeft(self: ?*anyopaque, v: ?*anyopaque) QtC.QJsonArray {
        return qtc.QJsonArray_OperatorShiftLeft(@ptrCast(self), @ptrCast(v));
    }

    pub fn PushBack(self: ?*anyopaque, t: ?*anyopaque) void {
        qtc.QJsonArray_PushBack(@ptrCast(self), @ptrCast(t));
    }

    pub fn PushFront(self: ?*anyopaque, t: ?*anyopaque) void {
        qtc.QJsonArray_PushFront(@ptrCast(self), @ptrCast(t));
    }

    pub fn PopFront(self: ?*anyopaque, ) void {
        qtc.QJsonArray_PopFront(@ptrCast(self));
    }

    pub fn PopBack(self: ?*anyopaque, ) void {
        qtc.QJsonArray_PopBack(@ptrCast(self));
    }

    pub fn Empty(self: ?*anyopaque, ) bool {
        return qtc.QJsonArray_Empty(@ptrCast(self));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QJsonArray_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qjsonarray-iterator.html
pub const qjsonarray__iterator = struct {

    /// New constructs a new QJsonArray::iterator object.
    pub fn New(other: ?*anyopaque) QtC.QJsonArray__iterator {
        return qtc.QJsonArray__iterator_new(@ptrCast(other));
    }


    /// New2 constructs a new QJsonArray::iterator object.
    pub fn New2() QtC.QJsonArray__iterator {
        return qtc.QJsonArray__iterator_new2();
    }


    /// New3 constructs a new QJsonArray::iterator object.
    pub fn New3(array: ?*anyopaque, index: i64) QtC.QJsonArray__iterator {
        return qtc.QJsonArray__iterator_new3(@ptrCast(array), index);
    }


    /// New4 constructs a new QJsonArray::iterator object.
    pub fn New4(other: ?*anyopaque) QtC.QJsonArray__iterator {
        return qtc.QJsonArray__iterator_new4(@ptrCast(other));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QJsonArray__iterator_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorPlusPlus(self: ?*anyopaque, ) It::value_type {
        return @ptrCast(qtc.QJsonArray__iterator_OperatorPlusPlus(@ptrCast(self)));
    }

    pub fn OperatorPlusPlus2(self: ?*anyopaque, param1: i32) It::value_type {
        return @ptrCast(qtc.QJsonArray__iterator_OperatorPlusPlus2(@ptrCast(self), param1));
    }

    pub fn OperatorMinusMinus(self: ?*anyopaque, ) It::value_type {
        return @ptrCast(qtc.QJsonArray__iterator_OperatorMinusMinus(@ptrCast(self)));
    }

    pub fn OperatorMinusMinus2(self: ?*anyopaque, param1: i32) It::value_type {
        return @ptrCast(qtc.QJsonArray__iterator_OperatorMinusMinus2(@ptrCast(self), param1));
    }

    pub fn OperatorPlusAssign(self: ?*anyopaque, j: i64) It::value_type {
        return @ptrCast(qtc.QJsonArray__iterator_OperatorPlusAssign(@ptrCast(self), j));
    }

    pub fn OperatorMinusAssign(self: ?*anyopaque, j: i64) It::value_type {
        return @ptrCast(qtc.QJsonArray__iterator_OperatorMinusAssign(@ptrCast(self), j));
    }

    pub fn OperatorPlus(self: ?*anyopaque, j: i64) It::value_type {
        return @ptrCast(qtc.QJsonArray__iterator_OperatorPlus(@ptrCast(self), j));
    }

    pub fn OperatorMinus(self: ?*anyopaque, j: i64) It::value_type {
        return @ptrCast(qtc.QJsonArray__iterator_OperatorMinus(@ptrCast(self), j));
    }

    pub fn OperatorMinus2(self: ?*anyopaque, j: QtC.QJsonArray__iterator) i64 {
        return qtc.QJsonArray__iterator_OperatorMinus2(@ptrCast(self), @ptrCast(j));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QJsonArray__iterator_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/qjsonarray-const_iterator.html
pub const qjsonarray__const_iterator = struct {

    /// New constructs a new QJsonArray::const_iterator object.
    pub fn New(other: ?*anyopaque) QtC.QJsonArray__const_iterator {
        return qtc.QJsonArray__const_iterator_new(@ptrCast(other));
    }


    /// New2 constructs a new QJsonArray::const_iterator object.
    pub fn New2() QtC.QJsonArray__const_iterator {
        return qtc.QJsonArray__const_iterator_new2();
    }


    /// New3 constructs a new QJsonArray::const_iterator object.
    pub fn New3(array: ?*anyopaque, index: i64) QtC.QJsonArray__const_iterator {
        return qtc.QJsonArray__const_iterator_new3(@ptrCast(array), index);
    }


    /// New4 constructs a new QJsonArray::const_iterator object.
    pub fn New4(o: *const It::value_type) QtC.QJsonArray__const_iterator {
        return qtc.QJsonArray__const_iterator_new4(@ptrCast(o));
    }


    /// New5 constructs a new QJsonArray::const_iterator object.
    pub fn New5(other: ?*anyopaque) QtC.QJsonArray__const_iterator {
        return qtc.QJsonArray__const_iterator_new5(@ptrCast(other));
    }


    pub fn OperatorAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QJsonArray__const_iterator_OperatorAssign(@ptrCast(self), @ptrCast(other));
    }

    pub fn OperatorPlusPlus(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonArray__const_iterator_OperatorPlusPlus(@ptrCast(self));
    }

    pub fn OperatorPlusPlus2(self: ?*anyopaque, param1: i32) QtC.QRect {
        return qtc.QJsonArray__const_iterator_OperatorPlusPlus2(@ptrCast(self), param1);
    }

    pub fn OperatorMinusMinus(self: ?*anyopaque, ) QtC.QRect {
        return qtc.QJsonArray__const_iterator_OperatorMinusMinus(@ptrCast(self));
    }

    pub fn OperatorMinusMinus2(self: ?*anyopaque, param1: i32) QtC.QRect {
        return qtc.QJsonArray__const_iterator_OperatorMinusMinus2(@ptrCast(self), param1);
    }

    pub fn OperatorPlusAssign(self: ?*anyopaque, j: i64) QtC.QRect {
        return qtc.QJsonArray__const_iterator_OperatorPlusAssign(@ptrCast(self), j);
    }

    pub fn OperatorMinusAssign(self: ?*anyopaque, j: i64) QtC.QRect {
        return qtc.QJsonArray__const_iterator_OperatorMinusAssign(@ptrCast(self), j);
    }

    pub fn OperatorPlus(self: ?*anyopaque, j: i64) QtC.QRect {
        return qtc.QJsonArray__const_iterator_OperatorPlus(@ptrCast(self), j);
    }

    pub fn OperatorMinus(self: ?*anyopaque, j: i64) QtC.QRect {
        return qtc.QJsonArray__const_iterator_OperatorMinus(@ptrCast(self), j);
    }

    pub fn OperatorMinus2(self: ?*anyopaque, j: QtC.QJsonArray__const_iterator) i64 {
        return qtc.QJsonArray__const_iterator_OperatorMinus2(@ptrCast(self), @ptrCast(j));
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QJsonArray__const_iterator_Delete(@ptrCast(self));
    }
};
